perma_web/perma/tests/test_link_count_caching.py

```diff
@@ -1,67 +1,52 @@
-from perma.models import Link, LinkUser
+def test_link_count_regular_user(link_user, link_factory):
+    """ We do some link count tallying on save """
+    link_count = link_user.link_count
+    link = link_factory(created_by=link_user, submitted_url="http://example.com")
+    link_user.refresh_from_db()
+    assert link_count + 1 == link_user.link_count
 
-from .utils import PermaTestCase
+    link.safe_delete()
+    link.save()
 
-class LinkCountCachingTestCase(PermaTestCase):
+    link_user.refresh_from_db()
+    assert link_count == link_user.link_count
 
-    @classmethod
-    def setUpTestData(cls):
-        cls.regular_user = LinkUser.objects.get(pk=4)
-        cls.org_user = LinkUser.objects.get(pk=3)
-        cls.registrar_user = LinkUser.objects.get(pk=2)
+def test_link_count_for_orgs(organization, link_user, org_link_factory):
+    """ We do some link count tallying on save. Let's make sure
+    we're adjusting the counts on the orgs """
 
-    ### TESTS ###
+    org_to_which_user_belongs = organization
+    link_user.organizations.set([org_to_which_user_belongs])
+    link_count = org_to_which_user_belongs.link_count
+    link = org_link_factory(created_by=link_user, submitted_url="http://example.com", organization=org_to_which_user_belongs)
+    link.save()
 
-    def test_link_count_regular_user(self):
-        """ We do some link count tallying on save """
+    org_to_which_user_belongs.refresh_from_db()
+    assert link_count + 1 == org_to_which_user_belongs.link_count
 
-        link_count = self.regular_user.link_count
-        link = Link(created_by=self.regular_user, submitted_url="http://example.com")
-        link.save()
+    link.safe_delete()
+    link.save()
 
-        self.regular_user.refresh_from_db()
-        self.assertEqual(link_count + 1, self.regular_user.link_count)
+    org_to_which_user_belongs.refresh_from_db()
+    assert link_count == org_to_which_user_belongs.link_count
 
-        link.safe_delete()
-        link.save()
 
-        self.regular_user.refresh_from_db()
-        self.assertEqual(link_count, self.regular_user.link_count)
+def test_link_count_for_registrars(registrar, link_user_factory, organization_factory, org_link_factory):
```

Comment:
Similarly, I think you can simplify this by doing more setup in the fixtures. I think this wants to use just a `registrar_user` and a `link_factory`. 

You would need to arrange for the `RegistrarFactory` to generate its related orgs, using the same strategy as:
```
@register_factory
class SponsoredUserFactory(LinkUserFactory):

    sponsorships = factory.RelatedFactoryList(
        SponsorshipFactory,
        size=1,
        factory_related_name='user'
    )
```
Let me know if that doesn't make sense!

Reply:
As I understand it, the reason that works is because there's a `user` field in the `sponsorship` model. Since there's no `organization` field on the `registrar` model, I don't think you can do quite the same thing. I'm reading from this documentation: https://factoryboy.readthedocs.io/en/stable/recipes.html#reverse-dependencies-reverse-foreignkey
Am I misunderstanding?
I will still try to simplify.

Reply:
It works because there is a `registrar` `ForeignKey` field on the `Organization` model :-)

Example implementation:

```
@register_factory
class RegistrarWithOrgsFactory(RegistrarFactory):

    organizations = factory.RelatedFactoryList(
        'conftest.OrganizationFactory',
        size=3,
        factory_related_name='registrar'
    )

def test_cremona(registrar_with_orgs):
    import pdb; pdb.set_trace()
    assert True

> /perma/perma_web/api/tests/test_link_resource.py(847)test_cremona()
-> assert True
(Pdb) registrar_with_orgs.organizations.all()
<OrganizationQuerySet [<Organization: Dean-Daniel>, <Organization: Green Inc>, <Organization: Jones Inc>]>
(Pdb) for org in registrar_with_orgs.organizations.all(): print(org.registrar)
Johnson, Robinson and Adams
Johnson, Robinson and Adams
Johnson, Robinson and Adams
(Pdb) registrar_with_orgs.name
'Johnson, Robinson and Adams'
```

Make sense? Happy to chat more if not!

There's probably no need to make a whole separate `Factory` subclass; probably harmless to add it to the regular one, though I haven't tried it and haven't thought it through. 

A fun trick might be, working out how to optionally pass in how many orgs to create, with a default of one.

Reply:
I tried adding as a param like this:

```
@register_factory
class RegistrarFactory(DjangoModelFactory):
    class Meta:
        model = Registrar
        exclude = ('num_orgs',)

    name = factory.Faker('company')
    email = factory.Faker('company_email')
    website = factory.Faker('url')
    num_orgs = 1

    # Default to "approved" in the fixtures for convenience
    status = 'approved'

    organizations = factory.RelatedFactoryList(
        'conftest.OrganizationFactory',
        size=num_orgs,
        factory_related_name='registrar'
    )
```

Unsuccessfully.

Reply:
Ack! I should not have tempted fate and used the word "fun".

Here's what I expected would work, which is almost exactly what you came up with, except instead of looking straight at the value of `num_orgs`, it makes use of `SelfAttribute`, which is how Factory Boy knows to look and see if you passed in an overriding value for `num_orgs` when you called the factory:
```
@register_factory
class RegistrarFactory(DjangoModelFactory):
    class Meta:
        model = Registrar

    class Params:
        num_orgs = 1

    name = factory.Faker('company')
    email = factory.Faker('company_email')
    website = factory.Faker('url')

    # Default to "approved" in the fixtures for convenience
    status = 'approved'

    organizations = factory.RelatedFactoryList(
        'conftest.OrganizationFactory',
        size=factory.SelfAttribute('num_orgs'),
        factory_related_name='registrar'
    )
```
But!!!!!! Turns out there is an [open issue](https://github.com/FactoryBoy/factory_boy/issues/767) about exactly this: the thing I expected to work.... does not work. Which is no fun. I put a 👍 on the issue in question.

There are workarounds, but meh, let's just hope we never need them.

Sorry for the misadventure!